File: qa-automation/src/pages/UploadAnalysisPage.js
const BasePage = require('./BasePage');
const logger = require('../utils/Logger');

class UploadAnalysisPage extends BasePage {
    // ----------------------------------------------------------------------
    // Selectors
    // ----------------------------------------------------------------------
    get pageHeader() { return $('.page-title'); }
    
    // Patient Selection (Doctor Only)
    get selectPatient() { return $('.patient-selector select'); }
    get lockedOverlay() { return $('.locked-overlay'); }
    
    // Upload Zone
    get uploadZone() { return $('.upload-zone'); }
    get inputFile() { return $('input#file-upload'); }
    get btnBrowse() { return $('label[for="file-upload"]'); }
    
    // States
    get emptyState() { return $('.empty-state'); }
    get scanningState() { return $('.scanning-state'); }
    get resultsState() { return $('.results-state'); }
    
    // Results
    get imagePreview() { return $('.image-preview'); }
    get boundingBoxes() { return $$('.bounding-box'); }
    get findingsTable() { return $('.findings-table'); }
    get markdownReport() { return $('.gemini-primary-report'); }
    
    // Actions
    get btnSaveReport() { return $('.results-state button:nth-of-type(1)'); }
    get btnScanAnother() { return $('.results-state button:nth-of-type(2)'); }

    // ----------------------------------------------------------------------
    // Methods
    // ----------------------------------------------------------------------

    /**
     * Navigates directly to analysis (requires auth session)
     */
    async open() {
        logger.info('Navigating to Upload & Analysis Page');
        await super.open('analysis');
        await this.waitForElement(this.pageHeader);
    }

    /**
     * Selects a patient from the dropdown (Doctor)
     */
    async selectFirstPatient() {
        logger.info('Selecting first available patient from dropdown');
        await this.waitForElement(this.selectPatient);
        
        // Wait for options to populate (skipping the placeholder)
        await browser.waitUntil(async () => {
            const options = await this.selectPatient.$$('option');
            return options.length > 1;
        }, { timeout: 10000, timeoutMsg: 'Patient list failed to load' });
        
        await this.selectPatient.selectByIndex(1);
        // Wait for locked overlay to disappear
        await browser.waitUntil(async () => {
            return !(await this.lockedOverlay.isDisplayed());
        }, { timeout: 5000 });
    }

    /**
     * Uploads an image using WebdriverIO setValue on the hidden input
     * @param {string} localFilePath 
     */
    async uploadImage(localFilePath) {
        logger.info(`Uploading image from path: ${localFilePath}`);
        // In WebdriverIO, if the input is hidden, we use execute to unhide it or just set value directly
        await browser.execute(function() {
            document.getElementById('file-upload').style.display = 'block';
        });
        
        await this.inputFile.setValue(localFilePath);
    }

    /**
     * Waits for the AI analysis to complete
     */
    async waitForAnalysisCompletion() {
        logger.info('Waiting for AI analysis to complete');
        await browser.waitUntil(async () => {
            return await this.resultsState.isDisplayed();
        }, { timeout: 30000, timeoutMsg: 'AI Analysis timed out or failed' });
    }
}

module.exports = new UploadAnalysisPage();

File: qa-automation/appium/pages/upload_analysis_page.py
from .base_page import BaseMobilePage
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import logging

logger = logging.getLogger("AppiumPython")

class MobileUploadAnalysisPage(BaseMobilePage):
    # Selectors
    PAGE_HEADER = (By.CSS_SELECTOR, '.page-title')
    
    # Patient Selection
    SELECT_PATIENT = (By.CSS_SELECTOR, '.patient-selector select')
    LOCKED_OVERLAY = (By.CSS_SELECTOR, '.locked-overlay')
    
    # Upload zone actions
    BTN_TAKE_PHOTO = (By.CSS_SELECTOR, '.native-upload-zone button.btn-primary')
    BTN_CHOOSE_FROM_GALLERY = (By.CSS_SELECTOR, '.native-upload-zone button.btn-outline')
    
    # States
    EMPTY_STATE = (By.CSS_SELECTOR, '.empty-state')
    SCANNING_STATE = (By.CSS_SELECTOR, '.scanning-state')
    RESULTS_STATE = (By.CSS_SELECTOR, '.results-state')
    
    # Results
    IMAGE_PREVIEW = (By.CSS_SELECTOR, '.image-preview')
    BOUNDING_BOXES = (By.CSS_SELECTOR, '.bounding-box')
    FINDINGS_TABLE = (By.CSS_SELECTOR, '.findings-table')
    BTN_SAVE_REPORT = (By.CSS_SELECTOR, '.results-state button:nth-of-type(1)')
    BTN_SCAN_ANOTHER = (By.CSS_SELECTOR, '.results-state button:nth-of-type(2)')

    # Native Android Selectors (to use in NATIVE_APP context)
    NATIVE_SHUTTER_BTN = (By.XPATH, '//android.widget.ImageView[@content-desc="Shutter" or @resource-id="com.android.camera2:id/shutter_button" or @resource-id="com.android.camera:id/shutter_button" or @content-desc="Take picture"]')
    NATIVE_CAMERA_ACCEPT_BTN = (By.XPATH, '//android.widget.ImageView[@content-desc="Done" or @resource-id="com.android.camera2:id/done_button" or @text="OK" or @text="SAVE" or @resource-id="com.android.camera2:id/confirm_button"]')
    NATIVE_FIRST_GALLERY_ITEM = (By.XPATH, '//android.widget.ImageView[contains(@resource-id, "icon") or contains(@resource-id, "thumbnail") or @content-desc="Photo" or @index="0"]')

    def capture_image_from_native_camera(self):
        """Triggers the native camera and captures an image"""
        logger.info("Appium: Tapping native Take Photo button in WebView")
        self.click_element(*self.BTN_TAKE_PHOTO)
        
        # Handle native permission dialogs if they show up
        self.handle_permission_dialog_if_shown()
        
        # Switch to native context for camera capture
        self.switch_to_native_context()
        
        logger.info("Appium Native: Clicking camera shutter button")
        self.click_element(*self.NATIVE_SHUTTER_BTN)
        time.sleep(2)  # Wait for capture processing
        
        logger.info("Appium Native: Clicking accept/confirm picture button")
        self.click_element(*self.NATIVE_CAMERA_ACCEPT_BTN)
        
        # Switch back to WebView context
        self.switch_to_webview_context()

    def select_image_from_native_gallery(self):
        """Selects an image from the device's native gallery picker"""
        logger.info("Appium: Tapping native Choose from Gallery button in WebView")
        self.click_element(*self.BTN_CHOOSE_FROM_GALLERY)
        
        # Handle native permission dialogs if they show up
        self.handle_permission_dialog_if_shown()
        
        # Switch to native context
        self.switch_to_native_context()
        
        logger.info("Appium Native: Selecting first image thumbnail from native photo gallery")
        self.click_element(*self.NATIVE_FIRST_GALLERY_ITEM)
        
        # Switch back to WebView context
        self.switch_to_webview_context()

    def select_first_patient(self):
        """Selects the first patient from the patient dropdown list"""
        logger.info("Appium: Selecting patient from dropdown")
        element = self.wait_for_element(*self.SELECT_PATIENT)
        Select(element).select_by_index(1)
        
        # Wait for value to change
        WebDriverWait(self.driver, 10).until(
            lambda d: Select(d.find_element(*self.SELECT_PATIENT)).first_selected_option.get_attribute("value") != ""
        )
